### DISCUSSION TITLE
Async atoms completely broken when there's Angular components on the page

### DISCUSSION THREAD

**Discussion by Lesik** [pmndrs/jotai#1963] [Bug report]
## Summary

I stumbled upon an extremely nasty bug on a page where my React component, using Jotai async atoms, is used alongside on a page where there's also Angular components. I don't have Angular experience, but from my understanding [zone.js](https://www.npmjs.com/package/zone.js) is quite core to Angular's functionality and it patches a lot of browser built-ins, among others `window.fetch` and `window.Promise`.

![image](https://github.com/pmndrs/jotai/assets/2021487/a5869afa-3bb7-4f4a-b68f-bca74424c8e7)

Jotai's async atoms rely heavily on promises (native browser promises), and [as outlined here](https://stackoverflow.com/a/28133130), the check `promise instanceof Promise` fails if the promise has been tampered with. There are about a dozen such checks in Jotai source code:

![image](https://github.com/pmndrs/jotai/assets/2021487/1ae18164-760e-45ad-86f5-b0728fbe5b64)

The way this surfaces depends on how you use async atoms:

- If you wrap your app in Suspense, then the "fallback" component is never rendered, it always goes to the child component even if the promise is still pending, causing an exception when the child component is trying to access a field in the promise data.
- If you use `loadable` from Jotai utils, the data you get back is `{ state: "hasData", data: Promise { <state>: "pending } }`, which is obviously nonsensical.

Yes, from our point of view as React developers, clearly this is Angular's fault. Who patches global built-ins on `window` in 2023? This is some jQuery-era garbage. But given that zone.js seems to be quite core to Angular, I don't believe there is an easy way of fixing this from the Angular side. And since integrating components from different frameworks into one app is quite common nowadays (and we don't have a choice, as the company we're integrating with runs Angular), I believe it would be pragmatic to fix this in Jotai, which would mean replacing the `promise instanceof Promise` check with one of the other suggested methods of detecting a promise.

## Link to reproduction

To reproduce this, you need both a React + Jotai project, build that, then add that built component to a page with an Angular component on it. I don't think you can do that in CSB, or if yes, I wouldn't know how.

## Check List

Please do not ask questions in issues.

- [ ] I've already opened a [discussion](https://github.com/pmndrs/jotai/discussions) before opening this issue, or already discussed in other media.

=> No, I believe the bug to be critical enough to go in an issue directly.

Please include a minimal reproduction.

- [ ] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

=> See above

**Comment by dai-shi** [comment]
I'm not sure if there's a factory standard, but we could make it somehow so that a developer can replace `instanceof Promise`.

> No, I believe the bug to be critical enough to go in an issue directly.

Let's take time and discuss what we want. We use discussions for bug report too.

**Reply by Lesik** [reply]
Well according to the SO post linked above, the greatest common denominator between all promise implementations is that they have a `then` function, so we could check for `typeof thing?.then === 'function'`. I'm not sure what the benefit would be of leaving this up to the developer, as then the default behavior of Jotai would still be "wrong", causing the above hard-to-debug bug (it's not immediately clear that the problem stems from the promise, as the symptoms surface somewhere else), and confusing developers working with Jotai.

**Comment by dai-shi** [comment]
out of curiosity, why doesn't `instanceof Promise` work if zone.js patches `globalThis.Promise` in advance? I believe a simple case of `instanceof Promise` would work, no?

**Reply by wootra** [reply]
I have tested during weekend, and some of promise alternative libraries (even though they have .then method), instanceof Promise does not work. 
I have tested q and bluebird. 
I have confirmed that there is PromiseLike type which just confirm if it has .then method, and if it has .then method, it works with async/await operator. 
it was interesting experiment. 
I also tested what would happen if I extend from Promise, and it worked as expected with instanceof Promise. 
I think it is simply because that Promise alternatives' core is not extended from Promise (maybe it came out before V8 engine is released since native Promise class belong to V8 engine).

**Reply by dai-shi** [reply]
Thanks for your investigation.
Can you try #1967 and see how far it fixes??

**Reply by Lesik** [reply]
I have built Jotai on the commit from that MR and confirm that it works on an Angular page now. Would it be possible to release this fix soon, so we don't have to fork? Thanks!

I'm just wondering why only two places were replaced, while there are more instances of `instanceof Promise` in the code, how can you be sure those will not trip over a patched promise?

Thanks again for the quick fix.

**Reply by dai-shi** [reply]
> how can you be sure those will not trip over a patched promise?

For `store.ts`, because they are created with `new Promise()` internally. It's under our control.
I'll check other utils later.

> so we don't have to fork?

It will still take some time, please feel free to use the codesandbox build in the mean time.
https://ci.codesandbox.io/status/pmndrs/jotai/pr/1967
☝️ See "Local Install Instructions"

**Reply by Lesik** [reply]
That's really cool, I didn't know CSB can do that!

May I ask out of curiosity what you are planning to do with this PR still that it will take some time? :smile:

**Reply by dai-shi** [reply]
I'll be double checking if there's anything I missed.

**Comment by t6adev** [comment]
Closing this due to a lack of response from the owner for an extended period.
Please feel free to reopen if the issue persists.

**Reply by Lesik** [reply]
By owner, you mean me? Because there was no question, therefore nothing for me to respond to :smile:  But the issue was resolved in #1967, so can be closed indeed.
